Updated add_dataset_transform() to have flexible insertion point

growlix opened the pull request:
`utils.data.add_dataset_transform()` by default appends transformations to `dataset.tranform.transforms`. This is how algorithms such as `randaugment` and `augmix` insert themselves. Some datasets already have `to_tensor()` in the list of transforms (e.g. Cifar10), however `randaugment` and `augmix` need to be added *before* `to_tensor()`. This PR updates `add_dataset_transform()` to have an optional `location` argument that allows insertion before `to_tensor()`.

hanlint commented:
Nice! Do we care to build a more generalized system? e.g.

```
before=ToTensor(), after=None
```

and allow someone to say.. insert this before `Normalize`?

hanlint commented on composer/utils/data.py:
```diff
@@ -95,26 +95,38 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
     Args:
         dataset: A torchvision-like dataset
         transform: Function to be added to the dataset's collection of transforms
+        location [str]: Where to insert the transform in the sequence of transforms. "end"
+        will append to the end, "before_totensor" will insert before ToTensor().
```
```suggestion
        location (str): Where to insert the transform in the sequence of transforms. "end"
        will append to the end, "before_totensor" will insert before ToTensor(). Default: "end".
```

hanlint reviewed (commented):
LGTM, but defer to @coryMosaicML for a review.

ravi-mosaicml commented on composer/utils/data.py:
```diff
@@ -95,26 +95,38 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
 
     if not hasattr(dataset, "transform"):
         raise ValueError(f"Dataset of type {type(dataset)} has no attribute 'transform'. Expected TorchVision dataset.")
+    assert location in ["end", "before_totensor"]
```
Can you do an if check and raise a valueerror instead of using assert? Asserts should not be used for data validation

ravi-mosaicml commented:
> Nice! Do we care to build a more generalized system? e.g.
> 
> ```
> before=ToTensor(), after=None
> ```
> 
> and allow someone to say.. insert this before `Normalize`?

Let's leave this for now. But since this parameter would only have two values, could we make it a boolean - i.e. `is_tensor_transformation: bool = True`?

growlix commented:
Made generalized as per @hanlint's request and idempotent as per @ravi-mosaicml's request

hanlint commented on composer/utils/data.py:
```diff
@@ -95,26 +96,61 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
+            `pre_post` argument. Default = "end". If `location` is not in list of
+            transforms, will append to end.
+        pre_post (str), optional: Whether to insert before ("pre") or after ("post") the
+            transform string passed in 'location'. Ignored if transform = "start" or "end". Default = "pre".
```
```suggestion
            transform string passed in 'location'. Ignored if location = "start" or "end". Default = "pre".
```

hanlint commented on composer/utils/data.py:
```diff
@@ -95,26 +96,61 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
+            insert at index 0, a transform (e.g. torchvision.transforms.ToTensor) will
+            insert pre- or post- the first instance of `location`, determined by the
+            `pre_post` argument. Default = "end". If `location` is not in list of
+            transforms, will append to end.
```
Would it be safer to error out if `location` is not in the list?

growlix replied:
Not every dataset has the same list of transforms, nor do they all have `ToTensor()` in the list of transforms. An error would require the user to provide `location` as a hyperparameter for the algorithm, which in turn would require them to know which transforms are in their dataset. This seems like a bit much. Perhaps a `log.warning()` displaying the list of transforms before the function returns?

hanlint replied:
i'm ok with a warning, would prefer an error to avoid silent fail (e.g. was supposed to run before Normalize, but since no Normalize, it's now silently running at the end).

ravi-mosaicml replied:
I find this API a bit confusing. Thoughts about replace the `location` and `pre_post` with two lists of transforms?

1. `pre_transformtions: Transformations`: Any transformations that, if present on `dataloader.transform`, will come before this new transformation
2. `post_transformations: Transformations`: Any transformations that, if present on `dataloader.transform`, will come this new transformation

If this is not possible (i.e. a post_transform already appears before a pre_transform), then a ValueError would be raised.

ravi-mosaicml replied:
Alternatively, I think we may be over engineering this. Thoughts about a singular `is_tensor_transform: bool` argument? If it is True, then it should be inserted after ToTensor; if it is False, then it should be inserted before ToTensor. If there is no `ToTensor` present, then it's inserted at the end.

Do our algorithms depend on a specific ordering (other than for the tensors)?

hanlint replied:
I'm also OK with `before=ToTensor()` and/or `after=end` but maybe that makes the input validation a bit difficult. @growlix you have the design decision on the API, so make a decision based on these inputs and we'll go with it!

hanlint commented on composer/utils/data.py:
```diff
@@ -95,26 +96,61 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
-    elif hasattr(dataset.transform, "transforms"):  # transform is a Compose
-        dataset.transform.transforms.append(transform)
+    # Check if dataset.transform is of type transforms.Compose and ensure idempotency by not adding transform if it's already present
+    elif isinstance(dataset.transform,
```
This is becoming sufficiently complex that perhaps we want to add some simple testing around this.

ravi-mosaicml replied:
Also, can this line be broken into multiple lines?

hanlint reviewed (approved):
LGTM, although we probably want to add a few tests.

ravi-mosaicml commented on composer/utils/data.py:
```diff
@@ -93,26 +94,61 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
+        else:
+            insertion_index = len(dataset.transform.transforms)
+            for i, t in enumerate(dataset.transform.transforms):
+                if type(t) == location:
```
Curious when would this be true?

growlix replied:
Not sure if you mean the `else` (L136) or the `if` (L139). L136 and below execute when `location` is not `"start"` or `"end"` (expected usage is that `location` is a `torchvision.transform`). L139 is true when the item in the list of transforms is of the same type as `location`, e.g. when they're both `transforms.ToTensor`.

ravi-mosaicml replied:
I think I missed that `t` could be a string or transform.

ravi-mosaicml commented on composer/utils/data.py:
```diff
@@ -93,26 +94,61 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
 
 
-def add_dataset_transform(dataset, transform):
+def add_dataset_transform(dataset, transform, location="end", pre_post: str = "pre") -> Dataset:
```
type annotation for dataset, transform, location

growlix commented:
I thought it'd be nice to have a flexible function that can insert a transform at location relative to other transforms, but you might be right about over-engineering. For now all we need is to insert before or after `ToTensor()` and to insert the transform somewhere else (default to end, or even always to end) if `ToTensor()` is not present.

growlix commented on composer/utils/data.py:
```diff
@@ -93,27 +97,86 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
+
+    def added_transform_warning():
+        log.warning(
+            f"Transform {transform} added to dataset. Dataset now has the following transforms: {dataset.transform}")
```
How do I deal with all the 
```
Cannot access member "transform" for type "Dataset[Unknown]"
Member "transform" is unknown
```
errors from the static analysis?

hanlint replied:
Hmm guess the static typechecker doesn't understand the guard you have upfront.

I'm not a typing expert, but I think you would use `Protocol` to implement a new type that has the attribute `transform`, and use that instead of `torch.utils.Dataset`, e.g. see [PEP 544](https://www.python.org/dev/peps/pep-0544/). 

But maybe defer to @ravi-mosaicml on this one.

ravi-mosaicml replied:
Can you add an `assert isinstance(dataset, VisionDataset)`?

hanlint commented on composer/utils/data.py:
```diff
@@ -93,27 +97,86 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
 
     if not hasattr(dataset, "transform"):
-        raise ValueError(f"Dataset of type {type(dataset)} has no attribute 'transform'. Expected TorchVision dataset.")
+        raise AttributeError(textwrap.dedent(f"""Dataset must have attribute `transform`."""))
```
do we need the `textwrap.dedent` here and below when the errors are relatively short?

ravi-mosaicml replied:
Nope, textwrap.dedent is only for multiline errors.

ravi-mosaicml replied:
Can we use check to ensure the dataset is a subclass of `torchvision.datasets.VisionDataset` (https://pytorch.org/vision/main/generated/torchvision.datasets.VisionDataset.html)?

e.g.

```python3
if not isinstance(dataset, torchvision.datasets.VisionDataset):
    raise TyepError(...)
```

hanlint replied:
Since we want to support custom datasets that have the `tranforms` attribute but not be properly inheriting from `VisionDataset`, we may not want to prematurely narrow the checks here.

ravi-mosaicml replied:
Any custom dataset can inherit from `VisionDataset`. This function was designed around the structure of `VisionDataset`, so I think it should be limited to datasets of that type.

growlix replied:
I don't think so, because Composer's synthetic datasets are of type `torch.utils.data.Dataset`, which would make them incompatible with this. But I think it could be reasonable to ensure that the dataset is a subclass of `torch.utils.data.Dataset`, which iirc is the most generic `dataset` class.

ravi-mosaicml replied:
Correct, our synthetic datasets do not support torchvision transforms, since the yield tensors directly. See https://github.com/mosaicml/composer/issues/346

ravi-mosaicml commented on composer/utils/data.py:
```diff
@@ -93,27 +97,86 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
-    else:  # transform is some other basic transform, join using Compose
-        dataset.transform = transforms.Compose([dataset.transform, transform])
-
+        added_transform = added_transform_warning()
```
For readabiity, can you remove the helper function and split it into two lines? Makes it more readable what's happening.

```suggestion
        log.warn(LOG_MESSAGE)
        added_transform = True
```
Same thing in the other places...

growlix commented on composer/utils/data.py:
```diff
@@ -93,27 +97,86 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
+    def added_transform_warning():
+        log.warning(
+            f"Transform {transform} added to dataset. Dataset now has the following transforms: {dataset.transform}")
+        return 1
```
To set a flag for whether the transform was actually added, but probably best not to combine those functions.

growlix commented:
I simplified `add_dataset_transform()` such that the only placement-related parameter is `is_tensor_transform`, and the new transform is placed before/after `ToTensor()` if `ToTensor()` is present in `dataset.transform`, and at the end if not. I also added a `SyntheticPILDataset` class that's nearly identical to the `SyntheticBatchPairDataset`, except that it generates/returns PIL images instead of tensors.

Also tests.

growlix commented on composer/utils/data.py:
```diff
@@ -93,27 +97,86 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
+    elif not isinstance(dataset.transform, transforms.Compose):
+        # Transform is some other basic transform, join using Compose
+        # Ensure idempotency by not adding transform if it's already present
+        if not type(dataset.transform) == type(transform):
```
No worries!

ravi-mosaicml commented on composer/core/types.py:
```diff
@@ -164,3 +164,8 @@ class MemoryFormat(StringEnum):
     PRESERVE_FORMAT = "preserve_format"
+
+
+class DatasetWithTransform(Protocol):
```
Curious is this class being used? If not let's remove it?

ravi-mosaicml commented on composer/datasets/synthetic.py:
```diff
@@ -125,3 +128,104 @@ def __getitem__(self, idx: int):
+                 num_classes: Optional[int] = None,
+                 label_shape: Optional[Sequence[int]] = None,
+                 transform: Optional[Callable] = None):
+        self.total_dataset_size = total_dataset_size
```
Can you add a call to `super.__init__(root="", transform=transform)`

ravi-mosaicml commented on composer/utils/data.py:
```diff
@@ -93,26 +98,60 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
     Args:
-        dataset: A torchvision-like dataset
-        transform: Function to be added to the dataset's collection of transforms
+        dataset (torch.utils.data.Dataset): A torchvision-like dataset
```
Docstring 
```suggestion
        dataset (VisionDataset): A torchvision-like dataset
```

ravi-mosaicml commented on composer/utils/data.py:
```diff
@@ -93,26 +98,60 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
+        transform (Callable): Function to be added to the dataset's collection of
+            transforms
+        is_tensor_transform (bool): Whether `transform` acts on data of the type
+        torch.Tensor. If `True`, and torchvision.transforms.ToTensor() is present in
```
For sphinx to parse it properly, lines 113-117 need to be indented.

ravi-mosaicml replied:
For auto-highlighting of methods, use these sphinx tags like `:meth:`

The leading `~` says to only show the last part (i.e. `ToTensor`) in the docs rather than the whole name (i.e. `torchvision.transforms.ToTensor`)

```suggestion
        torch.Tensor. If ``True``, and :meth:`~torchvision.transforms.ToTensor` is present in
```

Not going to comment on all, but let's enclose these in double-ticks so they will be formatted as code. Ideally we should reference it using the `:meth:` notation but can address that in a separate PR

ravi-mosaicml commented on composer/utils/data.py:
```diff
@@ -93,26 +98,60 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
+        dataset (torch.utils.data.Dataset): A torchvision-like dataset
+        transform (Callable): Function to be added to the dataset's collection of
+            transforms
+        is_tensor_transform (bool): Whether `transform` acts on data of the type
```
RST uses double ticks for code literal blocks (I made this mistake wayy to many times, then the docs come out in italics :D) Not going to mark all, but please double check.

```suggestion
        is_tensor_transform (bool): Whether ``transform`` acts on data of the type
```

ravi-mosaicml commented on composer/utils/data.py:
```diff
@@ -93,26 +98,60 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
         The original dataset. The transform is added in-place.
     """
 
     if not hasattr(dataset, "transform"):
```
Since the type annotation is a `VisionDataset`, this hasattr check is no longer needed.

ravi-mosaicml commented on composer/utils/data.py:
```diff
@@ -93,26 +98,60 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
+            insertion_index += 1
+        dataset.transform.transforms.insert(insertion_index, transform)
+        log.warning(transform_added_logstring)
+    elif not isinstance(dataset.transform, transforms.Compose):
```
Let's keep the comment, it's helpful :) 

Also, should this be an else?

```suggestion
    else: # transform is some other basic transform, join using Compose
```

ravi-mosaicml commented on composer/utils/data.py:
```diff
@@ -93,26 +98,60 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
+        else:
+            dataset.transform = transforms.Compose([dataset.transform, transform])
+            log.warning(transform_added_logstring)
+    else:
```
I don't think this else would ever trigger?

ravi-mosaicml commented on docs/source/callbacks.rst:
```diff
@@ -9,7 +9,7 @@ Composer contains built-in callbacks, which can be added via the ``--callbacks``
 .. code-block::
 
-    python examples/run_mosaic_trainer.py -f my_model.yaml --callbacks lr_monitor grad_monitor
+    python examples/run_composer_trainer.py -f my_model.yaml --callbacks lr_monitor grad_monitor
```
Whoops, thanks!

ravi-mosaicml commented on composer/datasets/synthetic.py:
```diff
@@ -125,3 +128,104 @@ def __getitem__(self, idx: int):
             return self.input_data[idx], self.input_target[idx]
+
+
+class SyntheticPILDataset(VisionDataset):
```
Thanks for writing this! This pretty implements #346 :)

Jotting this down for another issue/PR, unless if you have the bandwidth to implement it tmrw :):

A lot of this logic seems very similar to the `SyntheticBatchPairDataset`.

Could the `SyntheticPILDataset` internally use a `SyntheticBatchPairDataset` to generate the tensors, and then wrap it in a PIL image (and apply the transformations) on `__getitem__`.? I think that would clean up the code.

growlix replied:
Oh wow, nice to have inadvertently resolved an outstanding issue :)
I'm not quite sure how to internally use a `SyntheticBatchPairDataset` to generate the data for `SyntheticPILDataset` because `SyntheticBatchPairDataset` generates the tensors upon the first call to `__getitem__`, and `SyntheticPILDataset` would need to override `__getitem__`. It might be possible to change `SyntheticBatchPairDataset` to generate the data on init, but the comment for `__getitem__` in `SyntheticBatchPairDataset` makes it seem as though the data need to be generated upon the first call to `__getitem__` so that they're on the correct gpu.

ravi-mosaicml reviewed (approved):
Overall LGTM! See comments.

Can you try a docs build and make sure the docstrings format as expected? I tried to comment on some of the rst syntax errors but probably missed some. I don't want this to hold up the PR, though, since we're going to do a heavy docs scrub next week.
